analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -74,5 +70,59 @@ public sealed record SymbolicValue
             Constraints.Any()
                 ? Constraints.Values.Select(x => x.ToString()).OrderBy(x => x).JoinStr(", ")
                 : "No constraints";
+
+        private static SymbolicValue AddOrReplaceConstraint(SymbolicValue baseValue, SymbolicConstraint constraint)
+        {
+            if (baseValue.HasConstraint(constraint))
+            {
+                return baseValue;
+            }
+
+            var constraintCount = baseValue.Constraints.Count;
+            if (constraintCount == 0 || (constraintCount == 1 && baseValue.Constraints.ContainsKey(constraint.GetType())))
+            {
+                return SingleConstraint(constraint);
+            }
+
+            return baseValue with { Constraints = baseValue.Constraints.SetItem(constraint.GetType(), constraint) };
+        }
+
+        private static SymbolicValue RemoveConstraint(SymbolicValue baseValue, SymbolicConstraint constraint) =>
+            baseValue.HasConstraint(constraint) ? RemoveConstraint(baseValue, constraint.GetType()) : baseValue;
+
+        private static SymbolicValue RemoveConstraint<T>(SymbolicValue baseValue) where T : SymbolicConstraint =>
+            baseValue.HasConstraint<T>() ? RemoveConstraint(baseValue, typeof(T)) : baseValue;
+
+        private static SymbolicValue RemoveConstraint(SymbolicValue baseValue, Type type)
+        {
+            var constraintCount = baseValue.Constraints.Count;
+            if (constraintCount == 1)
+            {
+                return Constraintless;
+            }
+
+            if (constraintCount == 2)
+            {
+                SymbolicConstraint otherConstraint = null;
+                foreach (var kvp in baseValue.Constraints)
+                {
+                    if (kvp.Key != type)
+                    {
+                        otherConstraint = kvp.Value;
+                    }
+                }
+                return SingleConstraint(otherConstraint);
+            }
+
+            return baseValue with { Constraints = baseValue.Constraints.Remove(type) };
+        }
+
+        private static SymbolicValue SingleConstraint(SymbolicConstraint constraint) =>
+            constraint.Kind switch
```

Comment:
I still think the concurrent dictionary makes sense here as the penalty is small because the read operation on a concurrent dictionary is lock-free:

> (Read operations on the dictionary are performed in a lock-free manner.)

[Source](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2)

Reply:
Considering we'll have the concurrent dictionary, can we take it one step further to cache combinations too? Simple `OrderBy(x => x).JoinStr(", ", x => x.ToString())` should give a good-enough key

Basically for `True, NotNull`, to have a reused symbolic value. Or `LockHeld, NotNull` and so on.

Reply:
From looking at the other methods, it could simplify their implementation quite a lot too...

Reply:
> `OrderBy(x => x).JoinStr(", ", x => x.ToString())`

The benchmark has shown, that `baseValue with { Constraints = baseValue.Constraints.SetItem(constraint.GetType(), constraint) }` is quite sheep after all (My first implementation for one of the "RemoveConstraint" cases was more expensive with caching runtime and allocation-wise). I'm pretty certain that `OrderBy(x => x).JoinStr(", ", x => x.ToString())` is more expensive than creating just a new SymbolicValue. That is why I started with the single constraint and the idea was to do the same for a pair. See the description of the PR for how this can be done.
Anything that goes beyond two constraints would be kept as is with `baseValue with { Constraints = baseValue.Constraints.SetItem(constraint.GetType(), constraint) }`.